### PR TITLE
fixes #6991 - API v2 - support unwrapped POST/PUT parameters even though it's not documented

### PR DESCRIPTION
**Pull request by isratrade**
The goal of this PR is the preempt any possible bugs from API consumers that try to POST/PUT unwrapped params.

### CODE REVIEW

**Comment by isratrade** [comment]
Note, changes to api/v2/hosts_controller are in a separate PR #1600

File: app/controllers/api/v2/usergroups_controller.rb [hunk 1]
```diff
@@ -1,6 +1,9 @@
 module Api
   module V2
     class UsergroupsController < V2::BaseController
+
+      wrap_parameters :usergroup, :include => (Usergroup.attribute_names + ['user_ids', 'usergroup_ids', 'role_ids'])
```
Comment (domcleal) [inline app/controllers/api/v2/usergroups_controller.rb:5]: I see you've put _ids attributes here, and taxonomy_ids above.  Wouldn't these need to go into every single controller that has these type of relationships?

Comment (isratrade) [inline app/controllers/api/v2/usergroups_controller.rb:5]: @domcleal, yes, as a preemptive steps, we could add _ids in all places, if we are fully supporting unwrapped POST/PUT parameters, even the non-documented parameters like _ids. If I remember correctly, I think certain _ids were added in this PR to make tests pass.

File: app/controllers/concerns/api/v2/taxonomies_controller.rb [hunk 2]
```diff
@@ -14,16 +13,17 @@ module Api::V2::TaxonomiesController
...
+      param :smart_proxy_ids, Array, :desc => (N_("Array of smart proxy IDs to associate.") + DESC_WARNING_IDS)
+      param :compute_resource_ids, Array, :desc => (N_("Array of compute resource IDs to associate.") + DESC_WARNING_IDS)
+      param :media_ids, Array, :desc => (N_("Array of media IDs to associate.") + DESC_WARNING_IDS)
+      param :config_template_ids, Array, :desc => (N_("Array of provisioning template IDs to associate.") + DESC_WARNING_IDS)
+      param :domain_ids, Array, :desc => (N_("Array of domain IDs to associate.") + DESC_WARNING_IDS)
+      param :realm_ids, Array, :desc => (N_("Array of realm IDs to associate.") + DESC_WARNING_IDS)
+      param :hostgroup_ids, Array, :desc => (N_("Array of host group IDs to associate.") + DESC_WARNING_IDS)
+      param :environment_ids, Array, :desc => (N_("Array of environment IDs to associate.") + DESC_WARNING_IDS)
+      param :subnet_ids, Array, :desc => (N_("Array of subnet IDs to associate.") + DESC_WARNING_IDS)
+      param :subnet_ids, Array, :desc => (N_("Array of subnet IDs to associate.") + DESC_WARNING_IDS)
```
Comment (tstrachota) [inline app/controllers/concerns/api/v2/taxonomies_controller.rb]: `subnet_ids` are defined twice. Typo?

File: app/controllers/api/v2/base_controller.rb [hunk 3]
```diff
@@ -19,9 +19,11 @@ class BaseController < Api::BaseController
         param_group :pagination, ::Api::V2::BaseController
       end
 
+      DESC_WARNING_IDS = N_(" WARNING: Existing associations may be deleted if they are not included in the array.")
+
       def_param_group :taxonomies do
-        param :location_ids, Array, :required => false, :desc => N_("REPLACE locations with given ids") if SETTINGS[:locations_enabled]
-        param :organization_ids, Array, :required => false, :desc => N_("REPLACE organizations with given ids.") if SETTINGS[:organizations_enabled]
+        param :location_ids, Array, :required => false, :desc => (N_("Array of location IDs to associate.") + DESC_WARNING_IDS) if SETTINGS[:locations_enabled]
+        param :organization_ids, Array, :required => false, :desc => (N_("Array of location IDs to associate.") + DESC_WARNING_IDS) if SETTINGS[:organizations_enabled]
```
Comment (tstrachota) [inline app/controllers/api/v2/base_controller.rb]: Copy-paste typo in the param description.

File: app/controllers/api/v2/usergroups_controller.rb (hunk 1 above)
Comment (tstrachota) [inline app/controllers/api/v2/usergroups_controller.rb:5]: We could use `Model.reflections` to get the association names automatically.

**Comment by tstrachota** [comment]
Some comments in the code. Apart from them it looks good and passed all my tests.

**Comment by isratrade** [comment]
@tstrachota, updated PR per your comments.

**Comment by isratrade** [comment]
@tstrachota, rebased after merge conflicts.

**Comment by isratrade** [comment]
@eLobato, @ohadlevy, I think this is ready for merge. It fixes a BZ entered by Katello

**Comment by ohadlevy** [comment]
[test]

**Comment by isratrade** [comment]
@eLobato, @ohadlevy, ping. This fixes a BZ entered by Katello and will little affect on customers that follow the documentation using wrapped nodes.

**Comment by shlomizadok** [comment]
Tested. seems like it is ready to merge, though I dislike the unwrapped idea generally (+ this will kill us on Rails 4 + strong params)

**Comment by isratrade** [comment]
ping?

**Comment by isratrade** [comment]
@ohadlevy, @bkearney, rebased

**Comment by isratrade** [comment]
I recommend to close this PR.  I don't think we need to create a way for things to work (rather than throwing an error) if a user does _not_ according to the documentation. This is essentially what adding "wrapped_parameters" method in each controller allows.

**Pull request closed without merging**